src/video-grid/VideoGrid.tsx

```diff
@@ -692,38 +782,53 @@ interface DragTileData {
   y: number;
 }
 
-interface ChildrenProperties extends ReactDOMAttributes {
-  key: Key;
-  style: {
-    scale: SpringValue<number>;
-    opacity: SpringValue<number>;
-    boxShadow: Interpolation<number, string>;
-  };
-  width: number;
-  height: number;
-  item: TileDescriptor;
-  [index: string]: unknown;
+export interface ChildrenProperties<T> {
+  ref: Ref<HTMLElement>;
+  style: ComponentProps<typeof animated.div>["style"];
+  /**
+   * The width this tile will have once its animations have settled.
+   */
+  targetWidth: number;
+  /**
+   * The height this tile will have once its animations have settled.
+   */
+  targetHeight: number;
+  data: T;
 }
 
-interface VideoGridProps {
-  items: TileDescriptor[];
+export interface VideoGridProps<T> {
+  items: TileDescriptor<T>[];
   layout: Layout;
-  disableAnimations?: boolean;
-  children: (props: ChildrenProperties) => React.ReactNode;
+  disableAnimations: boolean;
+  children: (props: ChildrenProperties<T>) => React.ReactNode;
 }
 
-export function VideoGrid({
+// Represents something that should get a tile on the layout,
+// ie. a user's video feed or a screen share feed.
+export interface TileDescriptor<T> {
+  id: string;
+  focused: boolean;
+  isPresenter: boolean;
+  isSpeaker: boolean;
+  hasVideo: boolean;
+  local: boolean;
+  largeBaseSize: boolean;
+  placeNear?: string;
```

Comment:
First of all, thanks for the changes that bring new behavior to the grid layout, that was faster than expected 🙂 I think since we don't have to rush anymore, this seems like a good opportunity to discuss things that we could improve!

---

This change in the `TileDescriptor` (and the corresponding changes in the code that depends on it) seems to introduce a lot of complexity.  Originally (after the first refactoring that I did by generalizing the descriptor and removing unnecessary parts), we only had 2 boolean properties to describe a single tile: `focused` and `local` that contained 4 bits of information. Although `local` felt out of place (i.e. the layout/grid was expected to know what the "locality of participant" is and it was mostly used to control the layout in special cases of 1:1 calls, the complexity was still low (knowing that there are 4 potential states for the tile [although it was actually only 3] was bearable). Now we have 6 flags (**64 bits of information!**) along with the optional `placeNear`. So alone the flags allow a user to create **64 different combinations** of a tile description. This seems like much more state possibilities than we actually want to have 😉

The second problem is the encapsulation of logic. I think that in order to create a simple and reliable video grid layout logic, we could think of a grid layout as a separate generic component that exists in a separate library and does not know much about the type of content that is being rendered inside it, it simply accepts some input (the array of description of the tiles) and arranges them. Currently, we also have generic fields (`focused`) mixed with conferencing-specific fields that explicitly convey their meaning by their names (`hasVideo`, `isSpeaker`, etc). Moreover, the values of the flags are set in an unobvious way where the values overlap with each other in terms of meaning, e.g.:

```typescript
        const userMediaTile = {
          id,
          focused: false,
          isPresenter: sfuParticipant.isScreenShareEnabled,
          isSpeaker:
            (sfuParticipant.isSpeaking || hadSpokedInTime) &&
            !sfuParticipant.isLocal,
          hasVideo: sfuParticipant.isCameraEnabled,
          local: sfuParticipant.isLocal,
          largeBaseSize: false,
          data: {
            member,
            sfuParticipant,
            content: TileContent.UserMedia,
          },
        };
```

Note that `local` (`isLocal`) is set explicitly (i.e. the video grid knows when the field is local), but at the same time `isLocal` appears in the `isSpeaker` calculation which is counterintuitive - if the video grid knows that the tile is local, why does it expect that for calculating of certain other fields (`isSpeaker` in this case), the caller must include `local` in calculation? (it defeats the purpose of passing `local` to the video grid in the first place). Also note that `focused` is always set to `false` which also changes the meaning of what `focused` used to mean. It seems like it's easy to create a combination of flags that do not correspond to any set of the actual GUI state of a grid (i.e. to create a description that is "not supposed to be used").

I've shortly checked how these values are used and it seems like they are used to create some sort of a ranking system. Also, the meaning of some flags seems to change depending on the layout being used.

**One of the ways** to overcome the aforementioned issues is to implement the ranking explicitly - to pass a single integer value `priority` (or `rank`, or `importance` if you will) instead of a set of flags. This would allow the caller to specify the importance of a tile and to change the implementation of who is important and who isn't without ever changing the implementation logic of the video grid. This also allows generalizing the video grid algorithm to work for a more generic case. It also separates two concerns well. Also, it will be much more natural for the caller to describe the importance of the tiles as this will be done *explicitly* (instead of expecting that the video grid _knows_ that e.g. screen sharing tiles are more important than the rest of the tiles).

Then, calculating the importance of a tile may be as compact and elegant as something like this:

```rust
let importances = participants.map(|p| {
    let mut importance = 0;
    for criteria in &[p.isSpeaking, p.isScreensharing, p.hasCameraEnabled] {
        if criteria {
            importance += 1;
        }
    }
    return importance
});
```

If we discover that there are actually special cases that change the rendering so much that it does not fit the concept well, the better choice may be to create different layouts for them as different layouts may require different algorithms and hence different input parameters.